### PR TITLE
[WIP] Add tests

### LINKED ISSUE #7: Add tests
Tests would be nice to have, but how to implement them?

### PR DESCRIPTION
**Pull request by YurySolovyov**
I thought it would be nice to start a PR right away, so we can discuss everything as we go.

Some notes:
- fixed old image thumb, we should really find more reliable image with persistent url
- split up interface creators
- I started with jasmine as I thought it has pretty lightweight setup and everything I need ATM.

Feel free to leave a comment or commit some stuff as I will squash everything at the end.

Fixes #7

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -4,7 +4,7 @@
   "description": "Node.js implementation for the MPRIS D-Bus Interface Specification to create a mediaplayer service",
   "main": "index.js",
   "scripts": {
-    "test": "echo \"Error: no test specified\" && exit 1"
+    "test": "./node_modules/.bin/jasmine"
```
Comment (emersion) [inline package.json]: You can just write `jasmine`, NPM will expand it to `node_modules/.bin/jasmine` automatically.

**Comment by emersion** [comment]
Looks very good! We can also get rid of all the ugly `_create*Interface` methods by using `Symbol()` and use ES6 since this thing will only be used in Node.

Also, in the tests what happens if `done` is never called? Is there a timeout?

**Comment by YurySolovyov** [comment]
Some questions:
- do you have any particular minimal node version for this? I mean if you are ok with targeting node 4 LTS, we could use some ES6 here like const/let, classes, Promises and so on.
- I wonder if we can set up a CI to run the tests, though I'm not sure if we will be able to provide all dbus deps for it.

**Comment by YurySolovyov** [comment]
> Also, in the tests what happens if done is never called? Is there a timeout?

Yes

**Comment by emersion** [comment]
> do you have any particular minimal node version for this? I mean if you are ok with targeting node 4 LTS, we could use some ES6 here like const/let, classes, Promises and so on.

I think just targeting LTS will be fine. This module was written a _long_ time ago, and needs ES6.

> I wonder if we can set up a CI to run the tests, though I'm not sure if we will be able to provide all dbus deps for it.

I think this can work. Will try to do it.

File: .travis.yml [hunk 2]
```diff
@@ -3,3 +3,10 @@ node_js:
   - "6"
   - "5"
   - "4"
+env:
+  - CC=gcc-4.8 CXX=g++-4.8
+addons:
+  apt:
+    packages:
+      - gcc-4.8
+      - g++-4.8
```
Comment (YurySolovyov) [inline .travis.yml]: I guess this needs everything from https://github.com/Shouqun/node-dbus#dependencies plus `libexpat1-dev`

**Comment by YurySolovyov** [comment]
Comment got hidden: 
> I guess this needs everything from https://github.com/Shouqun/node-dbus#dependencies plus `libexpat1-dev`

**Comment by emersion** [comment]
Note: if we want to rollback to precise on travis, we need:
* Install expat from source (v2.0 is missing pkgconfig file, need v2.1): see https://wayland.freedesktop.org/ubuntu12.04.html
* Use a newer gcc adn g++:

```yaml
env:
  - CC=gcc-5 CXX=g++-5
addons:
  apt:
    sources:
      - ubuntu-toolchain-r-test
    packages:
      - gcc-5
      - g++-5
```

**Comment by emersion** [comment]
Yay! The build now works :D

File: spec/helpers/helpers.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+  },
+  playername: () => {
+    return 'test' + (Math.random() * 1000 | 0).toString();
+  },
+  servicename: (player) => {
+    return 'org.mpris.MediaPlayer2.' + player;
+  },
+  waitForEvent: (player, event) => {
+    return new Promise((resolve) => {
+      player.on(event, resolve);
```
Comment (emersion) [inline spec/helpers/helpers.js]: We should use `.once()` here to do not end up with a large number of listeners.

**Comment by YurySolovyov** [comment]
> Looks very good! We can also get rid of all the ugly _create*Interface methods by using Symbol()

TBH, I don't like using Symbol. How about just making them classes and using them like so:
```js
        
this.interfaces = {
  root: new RootInterface()
};
if (this.supportedInterfaces.includes('player')) {
	this.interfaces.player = new PlayerInterface();
}
if (this.supportedInterfaces.include('trackList')) {
	this.interfaces.trackList = new TrackListInterface();
}
if (this.supportedInterfaces.include('playlists')) {
	this.interfaces.playlists = new PlaylistsInterface();
}
```

File: spec/player-interface-spec.js [hunk 4]
```diff
@@ -67,21 +84,38 @@ describe('player interface', () => {
...
+
+  });
+
+  it('should emit signals on the bus that correspond to method calls', (done) => {
+
+    signals.reduce((promise, signal) => {
+
+      return promise.then(() => {
+
+        return helpers.getInterfaceAsync(service, objectpath, 'org.mpris.MediaPlayer2.Player').then(obj => {
```
Comment (emersion) [inline spec/player-interface-spec.js]: Is it possible to get the interface outside the `reduce`?

Comment (YurySolovyov) [inline spec/player-interface-spec.js]: good catch, will take a look

**Comment by YurySolovyov** [comment]
@emersion any news here?

**Comment by emersion** [comment]
Rebased all commits (but I should have just merged, too late :( ) with the latest fix in master. I get this:

```
Started
...*process 13125: arguments to dbus_message_iter_append_basic() were incorrect, assertion "_dbus_check_is_valid_path (*string_p)" failed in file dbus-message.c line 2758.
This is normally a bug in some application using the D-Bus library.
  D-Bus not built with -rdynamic so unable to print a backtrace
fish: “npm run test” terminated by signal SIGABRT (Abort)
```

**Comment by YurySolovyov** [comment]
You might want to show up in gitter :)

**Comment by emersion** [comment]
Superseded by https://github.com/emersion/mpris-service/pull/15

**Pull request closed without merging**